jax/lax/lax.py

```diff
@@ -2558,18 +2558,33 @@ def _reshape_impl(operand, new_sizes, dimensions, old_sizes):
       aval = ShapedArray(new_sizes, operand.dtype)
       lazy_expr = lazy.broadcast(operand._lazy_expr, new_sizes, bcast_dims)
       return xla.DeviceArray(aval, None, lazy_expr, operand.device_buffer)
-  if (type(operand) is pxla.ShardedDeviceArray and dimensions is None
-      and _is_axis_merge(old_sizes, new_sizes)):
-    aval = ShapedArray(new_sizes, operand.dtype)
-    return pxla.ChunkedDeviceArray(old_sizes[0], aval, operand.device_buffers)
-  elif (type(operand) is pxla.ChunkedDeviceArray and dimensions is None
-        and _is_axis_split(old_sizes, new_sizes)
-        and operand.axis_size == new_sizes[0]):
-    aval = ShapedArray(new_sizes, operand.dtype)
-    return pxla.ShardedDeviceArray(aval, operand.device_buffers)
-  else:
-    return xla.apply_primitive(reshape_p, operand, new_sizes=new_sizes,
-                               dimensions=dimensions, old_sizes=old_sizes)
+
+  if type(operand) is pxla.ShardedDeviceArray and dimensions is None:
+    # TODO(skye): the axis split/merge logic below assumes that
+    # ShardedDevicesArrays are always sharded across their leading axes. Remove
+    # this implicit constraint, especially if/when we add APIs that produce
+    # sharding across interior axes.
+    chunk_shape = operand.device_buffers[0].shape().dimensions()
+    if _is_axis_merge(old_sizes, new_sizes):
+      num_chunks, ragged = divmod(new_sizes[0], chunk_shape[0])
+      if not ragged:
+        aval = ShapedArray(new_sizes, operand.dtype)
+        # TODO(skye): deal with replication
```

Comment:
I think that it would be safest to replace these TODOs with NotImplemented exceptions.

Reply:
There's an assert in the SDA constructor that len(buffers) == len(indices).